CodeStatistic will now ignore hidden files:

Edouard-chin opened the pull request:
- Modify the default regex to not match hidden files

An `ArgumentError: invalid byte sequence in UTF-8` was raised if there was a binary file inside a directory. Those hidden files (with a `.rb`|`.js`... extension) were typically created by the filesystem.

| Q | A |
| --- | --- |
| Bug fix? | no, The exception will still be raised if a regular `.rb`, `.js`... contains binary text. This PR will be at least more convenient for the user, allowing him to run `rake stats` without first having to remove all those temporary files |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |

Fixes #22919

Edouard-chin committed:
CodeStatistic will now ignore hidden files:

- Modify the default regex to not match hidden files

thomasfedb commented:
LGTM :sparkles:

kaspth commented on railties/lib/rails/code_statistics.rb:
```diff
@@ -9,6 +9,8 @@ class CodeStatistics #:nodoc:
                 'Job tests',
                 'Integration tests']
 
+  DEFAULT_PATTERN = /^(?!\.).*?\.(rb|js|coffee|rake)$/
```
No need for extraction. Ruby deduplicates it when parsing the file the first time.

Edouard-chin replied:
:+1: Thanks good to know !

kaspth commented on railties/test/code_statistics_test.rb:
```diff
@@ -17,4 +17,18 @@ def teardown
+    File.write path, content
+    stats = CodeStatistics.new(['hidden file', @tmp_path])
+
+    assert_equal 0, stats.instance_variable_get("@statistics")['hidden file'].lines
```
This feels wrong. While I'm not the biggest fan of `assert_nothing_raised` it does feel like the lesser evil here, so let's go with that.

kaspth replied:
Also feel like we can get rid of some local variables by going with:

```ruby
File.write File.join(@tmp_path, '.example.rb'), <<-CODE
  def foo
    puts 'foo'
  end
CODE

assert_nothing_raised do
  CodeStatistics.new(['hidden file', @tmp_path])
end
```

Edouard-chin replied:
Nice, much cleaner!
Sorry for my ignorance but I'm curious to understand why an `assert_nothing_raised` should be tested here? Should the test focus on checking if the file was processed instead?

kaspth replied:
No worries :smile:

You're trying to remove a raised exception. I find `assert_nothing_raised` to better capture that than reaching into the statistics internal state 😁

kaspth commented:
Appreciate the diagram but adding "Fixes #22919" directly will make GitHub automatically close the original issue when this is merged.

Thanks for the pull request! :heart:

Edouard-chin commented:
Thanks for your helpful comment!

AFAIK "fixed tickets #22919" will not automatically close the associated issue

kaspth commented:
> AFAIK "fixed tickets #22919" will not automatically close the associated issue

It won't. That's why I'm suggesting `Fixes #22919`. Learn more here: https://github.com/blog/1506-closing-issues-via-pull-requests

Edouard-chin commented:
Modified PR according to your comments, and squashed my commits at the same time (sorry if I should not have). Modified description PR to close tickets when this get merged

Thanks :smiley:

kaspth commented:
You were right to squash, thanks! :smile:
